cli/context/store/store.go

```diff
@@ -0,0 +1,304 @@
+package store
+
+import (
+	"archive/tar"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"path"
+	"path/filepath"
+	"strings"
+)
+
+// Store provides a context store for easily remembering endpoints configuration
+type Store interface {
+	ListContexts() (map[string]ContextMetadata, error)
+	CreateOrUpdateContext(name string, meta ContextMetadata) error
+	RemoveContext(name string) error
+	GetContextMetadata(name string) (ContextMetadata, error)
+	ResetContextTLSMaterial(name string, data *ContextTLSData) error
+	ResetContextEndpointTLSMaterial(contextName string, endpointName string, data *EndpointTLSData) error
+	ListContextTLSFiles(name string) (map[string]EndpointFiles, error)
+	GetContextTLSData(contextName, endpointName, fileName string) ([]byte, error)
+}
+
+// Metadata contains metadata about an endpoint
+type Metadata map[string]interface{}
+
+// GetString retrieves a metadata value and checks it is a string
+func (m Metadata) GetString(key string) (string, bool) {
+	v, ok := m[key]
+	if !ok {
+		return "", false
+	}
+	s, ok := v.(string)
+	return s, ok
+}
+
+// GetBoolean retrieves a metadata value and checks it is a string
+func (m Metadata) GetBoolean(key string) (bool, bool) {
+	v, ok := m[key]
+	if !ok {
+		return false, false
+	}
+	b, ok := v.(bool)
+	return b, ok
+}
+
+// ContextMetadata contains metadata about a context and its endpoints
+type ContextMetadata struct {
+	Metadata  Metadata            `json:"metadata,omitempty"`
+	Endpoints map[string]Metadata `json:"endpoints,omitempty"`
```

Comment:
I was wondering why we didn't make this a `map[string]EndpointMeta` but I see there's two different versions of that (kube/docker), which is a pity (would've made things a bit more simple I guess 😞 

```go
type EndpointMeta struct {
	context.EndpointMetaBase
	APIVersion string
}

type EndpointMeta struct {
	context.EndpointMetaBase
	DefaultNamespace string
}
```


Wondering if we made this;

```go
Endpoints map[string]interface{}
```

Then we could do the conversion when _storing_, and store either `KubeEndpoint` or `DockerEndpoint` here 🤔

Reply:
:+1 It would greatly simplify Endpoint Marshalling/Unmarshalling

Reply:
Hmm, actually no, it wouldn't. But it would make it much cleaner (ContextMetadata would need to implement the json Marshal interface)